src/Microsoft.SymbolStore.Client/SymbolLocator.cs

```diff
@@ -0,0 +1,169 @@
+            SymbolServerResult result = await SearchServers((ISymbolServer server) => server.FindPEFileAsync(filename, timestamp, filesize));
+            if (result == null)
+                return null;
+
+            // The result may be a cab file instead of the actual result.  In this case, decompress it.
+            if (!result.Compressed)
+                return Cache.StorePEFile(result.Stream, filename, timestamp, filesize);
+            
+            CabConverter converter = new CabConverter(result.Stream);
+            using (MemoryStream stream = await converter.ConvertAsync())
+                return Cache.StorePEFile(stream, filename, timestamp, filesize);
+        }
+
+        private async Task<SymbolServerResult> SearchServers(Func<ISymbolServer, Task<SymbolServerResult>> getTask)
+        {
+            IEnumerable<Task<SymbolServerResult>> tasks = _symbolServers.Where(server => server.PreferThisServer).Select(server => getTask(server));
+            Task<SymbolServerResult> primaryResult = GetFirstNonNullResult(new List<Task<SymbolServerResult>>(tasks));
+
+            tasks = _symbolServers.Where(server => !server.PreferThisServer).Select(server => getTask(server));
+            Task<SymbolServerResult> secondaryResult = GetFirstNonNullResult(new List<Task<SymbolServerResult>>(tasks));
+
+            return await primaryResult ?? await secondaryResult;
+        }
+
+        private async Task<SymbolServerResult> FindPdbWorkerAsync(string pdbName, Guid guid, int age)
+        {
+            List<Task<SymbolServerResult>> processing = new List<Task<SymbolServerResult>>(_symbolServers.Length);
+            List<ISymbolServer> onDeck = new List<ISymbolServer>(_symbolServers.Length);
+
+            Task<SymbolServerResult> preferredResult = SearchServers((ISymbolServer server) => server.PreferThisServer, (ISymbolServer server) => server.FindPdbAsync(pdbName, guid, age));
+            Task<SymbolServerResult> secondaryResult = SearchServers((ISymbolServer server) => !server.PreferThisServer, (ISymbolServer server) => server.FindPdbAsync(pdbName, guid, age));
+
+            SymbolServerResult result = await preferredResult;
+            if (result != null)
+                return result;
+
+            return await secondaryResult;
+        }
+
+        private Task<SymbolServerResult> SearchServers(Func<ISymbolServer, bool> predicate, Func<ISymbolServer, Task<SymbolServerResult>> getTask)
+        {
+            var tasks = _symbolServers.Where(server => predicate(server)).Select(server => getTask(server));
+            return GetFirstNonNullResult(new List<Task<SymbolServerResult>>(tasks));
+        }
+        
+        async Task<T> GetFirstNonNullResult<T>(List<Task<T>> tasks) where T : class
+        {
+            while (tasks.Count > 0)
+            {
+                Task<T> task = await Task.WhenAny(tasks);
+
+                T result = task.Result;
+                if (result != null)
+                    return result;
```

Comment:
That's not quite how it works...  ISymbolServer only produces a Stream + metadata about that stream.  It does **not** read from the stream.  Our SearchServers/GetFirstNonNullResult will simply wait for the first symbol sever to say "I have this file, here is the stream".  If any other symbol sever produces a result, then we will discard that second stream and it will be GC'ed without ever having been read from.

Adding cancellation is tricky, since HttpWebRequest doesn't accept a cancellation token, and while the HttpWebRequest.Abort method can be used, the implementation is a little tricky.  Since we never read from the stream, we _should_ be ok to not cancel here...but if that turns out to not be the case we can certainly revisit this.